### PR TITLE
add identity field to AuthAccount

### LINKED ISSUE #423: Proving access to an authorized account
### Issue To Be Solved

A common use-case is proving access to an authorized account `AuthAccount`.

### Suggest A Solution

- Declare a new type `Identity` that cannot be constructed
- Add a new field `let identity: &Identity` to `AuthAccount`. The reference ensures the value cannot be stored

### PR DESCRIPTION
**Pull request by bjartek**
Closes #423 

## Description
Add an Identity type and a field to AuthAccount that retuns a reference to it. 

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Added appropriate labels

### CODE REVIEW

**Comment by bjartek** [comment]
@turbolent do you have any hints on how to go forward here, There is quite a lot boilerplate and I am not sure I have finished it all.

**Comment by joshuahannan** [comment]
I don't know if this makes sense. It encourages `msg.sender` access control list patterns that we have explicitly tried to avoid in Cadence.

**Comment by bjartek** [comment]
My usecase for this is to add context to events, since in some scenarios it is really hard to add "the other side" to an event.

**Comment by turbolent** [comment]
@joshuahannan Thank you for your feedback. Maybe we should propose such an addition through a FLIP?

**Comment by joshuahannan** [comment]
Maybe a FLIP would be good. Though I guess it wouldn't hurt to have a feature like this on the chance that it could be useful, but I worry that people would use it the wrong way. But that might not be the responsibility of the language to enforce and could be something that is just viewed as a best practice

**Comment by bluesign** [comment]
I agree with @joshuahannan in this case, but also I feel like we lost the war on msg.sender pattern with resource owner already. 

This will be more clear vs people abusing resource owner for sure. 

But eventually FLIP can be good to see other alternatives, community use cases etc.

File: docs/language/accounts.mdx [hunk 1]
```diff
@@ -167,6 +171,25 @@ to the `prepare` phase of the transaction.
   which discard their changes upon completion. 
   Attempting to use this function outside of a script will cause a type error. 
 
+## Identity
+Identity can only be obtained as a reference from an AuthAccount and has a single `address` field. 
+
+The main use case for Identity is to send it from a transaction into a contract method to know the address of who called it.
```
Comment (robert-e-davidson3) [inline docs/language/accounts.mdx:177]:
```suggestion
The main use case for Identity is to send it from a transaction into a contract method to know the caller's address.
```

File: docs/language/accounts.mdx [hunk 2]
```diff
@@ -167,6 +171,25 @@ to the `prepare` phase of the transaction.
...
+```cadence
+
+// example fun 
+pub fun someMethod(identity: &Identity) {
+   log("I was called by ".concat(identity.address))
+}
+
+transaction() {
+    prepare(signer: AuthAccount) {
+				someMethod(signer.identity)
```
Comment (robert-e-davidson3) [inline docs/language/accounts.mdx:188]: Whitespace shenanigans in a few places in this diff. Are they just an artifact of the medium or is the source code going to look weird?

File: runtime/sema/identity_type.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+import (
+	"github.com/onflow/cadence/runtime/common"
+)
+
+const IdentityTypeName = "Identity"
+const IdentityAddressField = "address"
+
+// IdentityType represents the Identity of an Auth Account
+// Identity can only be fetch as a reference from AuthAccount so it guarantees that the address in it comes from one of the signers.
```
Comment (robert-e-davidson3) [inline runtime/sema/identity_type.go:29]:
```suggestion
// Identity can only be fetched as a reference from AuthAccount so it guarantees that the address in it comes from one of the signers.
```

**Review by robert-e-davidson3 — APPROVED** [review]

**Comment by bjartek** [comment]
Sure

**Comment by turbolent** [comment]
The [FLIP proposing the addition of this feature ](https://github.com/onflow/flow/pull/945)has been rejected, so closing this implementation PR.

Thank you very much for your PR @bjartek, it sparked a very fruitful discussion, and hopefully an even better solution to the problem it tries to solve. 🙏

**Pull request closed without merging**